gcp_variant_transforms/libs/variant_partition.py

```diff
@@ -34,59 +43,295 @@
 
     Args:
       reference_name: reference name of the variant which is being partitioned
     Returns:
       An integer in the range of [0, _DEFAULT_NUM_PARTITIONS)
     """
-    partition = self._reference_name_to_partition_map.get(reference_name, None)
-    if partition is None:
+    indexer = self._by_ref_name.get(reference_name, None)
+    if indexer:
+      # In auto mode all partitions are whole chromosome. 
+      return indexer.get_index()
+    else:
       matched = _CHROMOSOME_NAME_REGEXP.match(reference_name)
       if matched:
         # First match the reference_name to the common formats.
         _, chr_no = matched.groups()
         chr_no = int(chr_no)
         if chr_no > 0 and chr_no <= _RESERVED_AUTO_PARTITIONS:
-          partition = chr_no - 1
-          self._reference_name_to_partition_map[reference_name] = partition
-          return partition
+          partition_index = chr_no - 1
+          self._by_ref_name[reference_name].add_index(partition_index)
+          return partition_index
       # If RegExp didn't match, we will find the hash of reference_name
       remaining_partitions = _DEFAULT_NUM_PARTITIONS - _RESERVED_AUTO_PARTITIONS
-      partition = (hash(reference_name) % remaining_partitions +
-                   _RESERVED_AUTO_PARTITIONS)
+      partition_index = (hash(reference_name) % remaining_partitions +
+                         _RESERVED_AUTO_PARTITIONS)
       # Save partition in _reference_name_to_partition dict for future lookups
-      self._reference_name_to_partition_map[reference_name] = partition
-    return partition
+      self._by_ref_name[reference_name].add_index(partition_index)
+      return partition_index
+
+  def has_residual_partition(self):
+    return self._residual_partition_index != -1
+
+  def should_flatten(self):
+    # In case we are in auto mode (ie no config) we flatten partitions and
+    # produce only one output table.
+    return not self._config_file_path_given
+
+  def get_residual_partition_index(self):
+    # In auto mode residual partition doesnt exist, return -1.
+    if not self._config_file_path_given:
+      return -1
+
+    # If residual partition is absent in the config file, we still need an index
+    # for all residual variants. In this case we add a dummy paritition indexed
+    # _num_partitions - 1 which will be ignored right after the partitioning.
+    if self.has_residual_partition():
+      return self._residual_partition_index
+    else:
+      return self._num_partitions - 1
+
+  def get_partition_name(self, index):
+    if self._config_file_path_given:
+      if index >= self._num_partitions or index < 0:
+        raise ValueError('Partition index is outside of expected range: {}'.
+                         format(index))
+      # Note, a '_' is automatically added, so output table name would be:
+      #   known_args.output_table + '_' + partition_name
+      return '_' + self._partition_names[index]
+    else:
```

Comment:
How about returning empty string for all else cases? This will make the  vcf_to_bq.py look cleaner (in the loop where we read the partition_name and call variant_to_bigquery.VariantToBigQuery.

Let me know if you think this is reasonable.